execution/state/events.go

```diff
@@ -1,64 +1,83 @@
 package state
 
 import (
+	"bytes"
 	"fmt"
 	"io"
 
 	"github.com/hyperledger/burrow/execution/exec"
 )
 
 func (ws *writeState) AddBlock(be *exec.BlockExecution) error {
-	tree, err := ws.forest.Writer(keys.Event.Prefix())
-	if err != nil {
-		return err
-	}
-	txHashTree, err := ws.forest.Writer(keys.TxHash.Prefix())
-	if err != nil {
-		return err
-	}
-	// Index transactions so they can be retrieved by their TxHash
-	for index, ev := range be.StreamEvents() {
-		// Store with prefix for scanning later
-		key := keys.Event.KeyNoPrefix(be.Height, uint64(index))
-		bs, err := ev.Encode()
-		if err != nil {
-			return err
-		}
-		// Set StreamEvent itself
-		tree.Set(key, bs)
+	txebs := make([]byte, 0)
+
+	for _, ev := range be.StreamEvents() {
 		if ev.BeginTx != nil {
+			val := &exec.TxExecutionKey{Height: be.Height, Offset: uint64(len(txebs))}
+			bs, err := val.Encode()
+			if err != nil {
+				return err
+			}
 			// Set reference to TxExecution
-			txHashTree.Set(keys.TxHash.KeyNoPrefix(ev.BeginTx.TxHeader.TxHash), key)
+			ws.plain.Set(keys.TxHash.Key(ev.BeginTx.TxHeader.TxHash), bs)
+		}
+
+		bs, err := cdc.MarshalBinaryLengthPrefixed(ev)
+		if err != nil {
+			return err
 		}
+
+		txebs = append(txebs, bs...)
```

Comment:
what is `txebs`? would be more appropriate as something like `evbs` or `streamEventBytes`